can/io/mf4.py

```diff
@@ -0,0 +1,392 @@
+        ("CAN_ErrorFrame.Dir", "<u1"),
+        ("CAN_ErrorFrame.EDL", "<u1"),
+        ("CAN_ErrorFrame.BRS", "<u1"),
+        ("CAN_ErrorFrame.ESI", "<u1"),
+    ]
+)
+
+RTR_DTYPE = np.dtype(
+    [
+        ("CAN_DataFrame.BusChannel", "<u1"),
+        ("CAN_DataFrame.ID", "<u4"),
+        ("CAN_DataFrame.IDE", "<u1"),
+        ("CAN_DataFrame.DLC", "<u1"),
+        ("CAN_DataFrame.Dir", "<u1"),
+    ]
+)
+
+FD_LEN2DLC = {12: 9, 16: 10, 20: 11, 24: 12, 32: 13, 48: 14, 64: 15}
+
+FD_DLC2LEN = {value: key for key, value in FD_LEN2DLC.items()}
+
+logger = logging.getLogger("can.io.mf4")
+
+
+class MF4Writer(BaseIOHandler, Listener):
+    """Logs CAN data to an ASAM Measurement Data File v4 (.mf4).
+
+    MF4Writer does not support append mode.
+
+    If a message has a timestamp smaller than the previous one or None,
+    it gets assigned the timestamp that was written for the last message.
+    It the first message does not have a timestamp, it is set to zero.
+    """
+
+    def __init__(self, file, database=None):
+        """
+        :param file: a path-like object or as file-like object to write to
+                        If this is a file-like object, is has to be opened in
+                        binary write mode, not text write mode.
+        :param database: optional path to a DBC or ARXML file that contains
+                            message description.
+        """
+        super().__init__(file, mode="r+b")
+        now = datetime.now()
+        self._mdf = MDF4()
+        self._mdf.header.start_time = now
+        self.last_timestamp = self._start_time = now.timestamp()
+
+        if database:
+            database = Path(database).resolve()
+            if database.exists():
+                data = database.read_bytes()
+                attachment = data, database.name
+            else:
+                attachment = None
+        else:
+            attachment = None
+
+        # standard frames group
+        self._mdf.append(
+            Signal(
+                name="CAN_DataFrame",
+                samples=np.array([], dtype=STD_DTYPE),
+                timestamps=np.array([], dtype="<f8"),
+                attachment=attachment,
+            )
+        )
+
+        # error frames group
+        self._mdf.append(
+            Signal(
+                name="CAN_ErrorFrame",
+                samples=np.array([], dtype=ERR_DTYPE),
+                timestamps=np.array([], dtype="<f8"),
+                attachment=attachment,
+            )
+        )
+
+        # remote frames group
+        self._mdf.append(
+            Signal(
+                name="CAN_DataFrame",
+                samples=np.array([], dtype=RTR_DTYPE),
+                timestamps=np.array([], dtype="<f8"),
+                attachment=attachment,
+            )
+        )
+
+        self._buffer = np.zeros(1, dtype=STD_DTYPE)
+        self._rtr_buffer = np.zeros(1, dtype=RTR_DTYPE)
+
+    def stop(self):
+        self._mdf.save(self.file, compression=2)
```

Comment:
Would it make sense to make this compression option configurable?

Reply:
Added compression argument to stop method